### PR TITLE
Cursor preferences

### LINKED ISSUE #1857: Hiding the red circle on the brush cursor
Hello,

I find that the size indicator on the brush tool can be very distracting while sketching out a rough animation.

Would it be possible to add an interface option and a shortcut to toggle it?

Thank you.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/55990257-hiding-the-red-circle-on-the-brush-cursor?utm_campaign=plugin&utm_content=tracker%2F33713530&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F33713530&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by manongjohn**
This PR allows users to set some cursor preferences on drawing tools.  This currently only affects cursors on the canvas and not places like xsheet/timeline or toolbars, etc.

The preferences found under the Tools tab in the Preferences window and include the following:

- [x] Use Simple Cursors

- Some people find some of the icons annoying.  This converts them to a +

- [x] Use Left-Handed Cursors

- For those lefties out there, this flips most cursors horizontally.

- [x] Show Cursor Size Outlines

- This PR fixes #1857 providing the option to show or hide the circle surrounding the cursor which normally indicates the area of effect or size of the stroke.  This can be toggled easily via the context menu when right-clicking the canvas.  A keyboard shortcut can be assigned to this action as well.

### CODE REVIEW

**Comment by manongjohn** [comment]
@artisteacher Thanks for the suggestion. I will move it there.

**Comment by konero** [comment]
It works! Very cool. 💪 

I hate to be that guy, but would it be too much work adding a combo box to allow users to choose the style for the `ToolCursor::PenCursor`? 🙏 

This isn't a big problem, but I did wonder about the size, it might just be me and having large screens but I always thought it was too small since I do most filling operations with a mouse instead of a stylus.

I made two cross cursors for testing based from other popular software. The images attatched here can be downloaded by right-click Save As.

The original one from OpenToonz (small).
brush

One based off TVPaint/PaintDotNet (large).
brush

And one a lot of other software uses (crosshair) like Clip Studio/FlipBook/SketchBook.
brush2

They all have a white border like the original does:

yoyo_cursor

cursor_opt_m

**Comment by manongjohn** [comment]
@konero You know...you're always making more work for me!  Lol

Just kidding, It's all good!

I was concerned about the size of the pen brush being too small as well and can add this option relatively easily unless @shun-iwasawa is ok with just replacing the pen brush with one of the bigger ones you've provided (hint, nudge, hint, hint, grin)

**Comment by gab3d** [comment]
to me the Crosshair one has the benefit of being noticeable, while at the same time, letting the user see (by not blocking) the exact area the tool will act upon.

File: toonz/sources/tnztools/cursormanager.cpp [hunk 1]
```diff
@@ -132,12 +146,75 @@ class CursorManager {  // singleton
...
+      if (it != m_cursors.end()) return it->second;
+    }
 
     int decorationsFlag = cursorType & ~(0xFF);
     int baseCursorType  = cursorType & 0xFF;
 
+    switch (baseCursorType) {
+    case ToolCursor::FxGadgetCursor:
+      decorationsFlag |= ToolCursor::Ex_FX;
+      break;
```
Comment (shun-iwasawa) [inline toonz/sources/tnztools/cursormanager.cpp]: Here you are trying to find specific tool type to add a decoration which cannot be generated by simple mirroring.
Instead, how about including such decoration information in the cursor flag like as follows?

```
-------------------- toonz/sources/include/tools/cursors.h --------------------
@@ -75,7 +75,7 @@ enum {
   MoveZCursor,
   ScaleGlobalCursor,
   ScaleHVCursor,
-  FxGadgetCursor,
+  FxGadgetCursorBase,
   RulerModifyCursor,
   RulerNewCursor,
 
@@ -92,7 +92,9 @@ enum {
   Ex_StyleLine        = 0x20000,
   Ex_StyleArea        = 0x40000,
   Ex_RGB              = 0x80000,
-  Ex_HV               = 0x100000
+  Ex_HV               = 0x100000,
+
+  FxGadgetCursor = FxGadgetCursorBase | Ex_FX
 };
 
 }  // namespace

------------------- toonz/sources/tnztools/cursormanager.cpp -------------------
@@ -34,7 +34,7 @@ const struct {
     {ToolCursor::MoveNSCursor, "move_ns", 15, 15, false},
     {ToolCursor::DisableCursor, "disable", 15, 15, false},
     {ToolCursor::MoveZCursor, "move_z_notext", 15, 15, false},
-    {ToolCursor::FxGadgetCursor, "edit_FX_notext", 11, 6, true},
+    {ToolCursor::FxGadgetCursorBase, "edit_FX_notext", 11, 6, true},
     {ToolCursor::FlipHCursor, "flip_h", 15, 15, false},
     {ToolCursor::FlipVCursor, "flip_v", 15, 15, false},
     {ToolCursor::IronCursor, "iron", 15, 15, true},
```

File: toonz/sources/tnztools/cursormanager.cpp [hunk 2]
```diff
@@ -17,64 +19,67 @@ const struct {
...
+    {ToolCursor::RotBottomRight, "rot_bottom_right", 15, 15, false},
+    {ToolCursor::RotBottomLeft, "rot_bottom_left", 15, 15, false},
+    {ToolCursor::RotateCursor, "rotate", 15, 19, true},
+    {ToolCursor::ScaleCursor, "scale", 15, 15, false},
+    {ToolCursor::ScaleInvCursor, "scale_inv", 15, 15, false},
+    {ToolCursor::ScaleHCursor, "scale_h", 15, 15, false},
+    {ToolCursor::ScaleVCursor, "scale_v", 15, 15, false},
+    {ToolCursor::EditFxCursor, "edit_FX_notext", 11, 6, true},
+    {ToolCursor::ScaleGlobalCursor, "scale_global", 15, 15, false},
+    {ToolCursor::ScaleHVCursor, "scale_hv", 15, 15, false},
```
[the remark this answers is not in the document]
Comment (manongjohn) [reply toonz/sources/tnztools/cursormanager.cpp]: Yes.  Thanks for catching

File: toonz/sources/tnztools/cursormanager.cpp (hunk 1 above)
Comment (manongjohn) [reply toonz/sources/tnztools/cursormanager.cpp]: Changes made. Thanks for the recommendation

**Comment by shun-iwasawa** [comment]
Jenkins

**Comment by shun-iwasawa** [comment]
Confirmed it works fine. Thank you @manongjohn for the improvement!
LGTM

**Pull request merged**